stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -1192,40 +1476,101 @@ impl NakamotoChainState {
         Ok(Some(receipt))
     }
 
+    /// Get the expected total burnchain tokens spent so far for a given block.
+    /// * if the block has a tenure-change tx, then this is the tx's sortition consensus hash's
+    /// snapshot's burn total (since the miner will have produced this tenure-change tx in reaction
+    /// to the arrival of this new sortition)
+    /// * otherwise, it's the highest processed tenure's sortition consensus hash's snapshot's burn
+    /// total.
+    ///
+    /// TODO: unit test
+    pub(crate) fn get_expected_burns<SH: SortitionHandle>(
+        sort_handle: &mut SH,
+        chainstate_conn: &Connection,
+        block: &NakamotoBlock,
+    ) -> Result<Option<u64>, ChainstateError> {
+        let target_ch = if let Some(tenure_payload) = block.get_tenure_tx_payload() {
+            tenure_payload.sortition_consensus_hash
+        } else if let Some(highest_tenure) =
+            Self::get_highest_nakamoto_tenure(chainstate_conn, sort_handle)?
+        {
+            highest_tenure.sortition_consensus_hash
+        } else {
+            // no nakamoto tenures yet, so this is the consensus hash of the canonical stacks tip
+            let (consensus_hash, _) =
+                SortitionDB::get_canonical_stacks_chain_tip_hash(sort_handle.sqlite())?;
+            consensus_hash
+        };
+
+        let Some(sn) = SortitionDB::get_block_snapshot_consensus(sort_handle.sqlite(), &target_ch)?
+        else {
+            warn!("Unacceptable Nakamoto block -- no sortition for tenure";
+                  "sortition_consensus_hash" => %target_ch
+            );
+            return Ok(None);
+        };
+        Ok(Some(sn.total_burn))
+    }
```

Comment:
This function seems strange to me -- the expected burns for a block are intrinsic to that block and its view of the burnchain. Querying the "highest known tenure" for that information may be correct as well, but that's only context dependent, and it requires multiple steps to reason about.

For example, if a block is mined in Tenure 1, and has a consensus_hash view corresponding to bitcoin block 2, then the total burn amount for that block should *always and forever* be the burn block total at bitcoin block 2. However, as this function is currently written, it would return the burn block total at bitcoin block 2 when first evaluated, but if someone called this function after subsequent tenure's are processed on the _same block_, they would get a different burn amount.

I understand that this function is only to be called while processing a newly arrived block, but this seems like very avoidable behavior:

```rust
let burn_view_ch = if let Some(tenure_payload) = block.get_tenure_tx_payload() {
      tenure_payload.sortition_consensus_hash
} else {
      // if there's no new tenure for this block, the burn total should be the same as its parent
     let parent = Self::get_block_header(chainstate_conn, &block.parent_block_id)?
          .unwrap_or_else(|| ChainstateError::NoSuchParent)?;
    // I think we should add a `.total_burn()` method to StacksHeaderInfo: both epoch2 and epoch3 headers have that information.
     return Ok(parent.total_burn()) 
};
let burn_view_sn = SortitionDB::get_block_snapshot_consensus(sort_handle.sqlite(), &target_ch)?
          .unwrap_or_else(|| ChainstateError::NoSuchParent)?;
Ok(burn_view_sn.total_burn)
```

Reply:
Sure, I can do that